### PR TITLE
Styles API

### PR DESCRIPTION
**Pull request by SomeHats**
Removes `propsForNextShape` and replaces it with the new styles API. 

Changes in here:
- New custom style example
- `setProp` is now `setStyle` and takes a `StyleProp` instead of a string
- `Editor.props` and `Editor.opacity` are now `Editor.sharedStyles` and `Editor.sharedOpacity`
    - They return an object that flags mixed vs shared types instead of using null to signal mixed types
    - `Editor.styles` returns a `SharedStyleMap` - keyed on `StyleProp` instead of `string`
- `StateNode.shapeType` is now the shape util rather than just a string. This lets us pull the styles from the shape type directly.
- `color` is no longer a core part of the editor set on the shape parent. Individual child shapes have to use color directly.
- `propsForNextShape` is now `stylesForNextShape`
- `InstanceRecordType` is created at runtime in the same way `ShapeRecordType` is. This is so it can pull style validators out of shape defs for `stylesForNextShape`
- Shape type are now defined by their props rather than having separate validators & type defs

### Change Type

- [x] `major` — Breaking change

### Test Plan

1. Big time regression testing around styles!
2. Check UI works as intended for all shape/style/tool combos

- [x] Unit Tests
- [ ] End to end tests

### Release Notes

-

### CODE REVIEW

File: apps/examples/src/16-custom-styles/CustomStylesExample.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+				shapes={shapes}
+				overrides={cardToolMenuItems}
+			>
+				<FilterStyleUi />
+			</Tldraw>
+		</div>
+	)
+}
+
+const FilterStyleUi = track(function FilterStyleUi() {
```
Comment (SomeHats) [inline apps/examples/src/16-custom-styles/CustomStylesExample.tsx:25]: this little bit of UI shows what using the new API more interactively is like

File: packages/editor/src/lib/constants.ts [hunk 2]
```diff
@@ -119,157 +119,41 @@ export const TEXT_PROPS = {
...
-
-/** @public */
-export const FONT_ALIGNMENT: Record<TLAlignType, string> = {
-	middle: 'center',
-	start: 'left',
-	end: 'right',
-}
-
-/** @public */
-export const STYLES: TLStyleCollections = {
```
Comment (SomeHats) [inline packages/editor/src/lib/constants.ts:169]: this lives in the ui now

File: packages/editor/src/lib/editor/Editor.ts [hunk 3]
```diff
@@ -3345,59 +3319,6 @@ export class Editor extends EventEmitter<TLEventMap> {
...
-	 * editor.getStrokeWidth('m')
-	 * ```
-	 *
-	 * @param id - The id of the size to get.
-	 * @public
-	 */
-	getStrokeWidth(id: TLSizeStyle['id']): number {
-		return this.sizes[id]
-	}
-
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:3400]: all this stuff is sunk into shape utils now

File: packages/editor/src/lib/editor/Editor.ts [hunk 4]
```diff
@@ -5792,16 +5708,17 @@ export class Editor extends EventEmitter<TLEventMap> {
...
+					// for shapes to register fonts for export.
+					const fontFromShape = util.getStyleIfExists(DefaultFontStyle, shape)
+					if (fontFromShape) {
+						if (fontsUsedInExport.has(fontFromShape)) {
+							font = fontsUsedInExport.get(fontFromShape)!
+						} else {
+							// For some reason these styles aren't present in the fake element
+							// so we need to get them from the real element
+							font = realContainerStyle.getPropertyValue(`--tl-font-${fontFromShape}`)
+							fontsUsedInExport.set(fontFromShape, font)
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:5718]: any ideas how we can avoid `Editor` knowing about `DefaultFontStyle` here?

File: packages/editor/src/lib/editor/Editor.ts [hunk 5]
```diff
@@ -7961,77 +7889,62 @@ export class Editor extends EventEmitter<TLEventMap> {
 					}
 
 					this.updateShapes(
-						shapesToUpdate.map((shape) => {
-							const props = { ...shape.props, [key]: value }
-							if (key === 'color' && 'labelColor' in props) {
-								props.labelColor = 'black'
-							}
-
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:7969]: i'm not 100% on what this logic was there for, but i've removed it as i don't see how `labelColor` could be anything other than `black` anyway. Any ideas?

File: packages/editor/src/lib/editor/Editor.ts [hunk 6]
```diff
@@ -7961,77 +7889,62 @@ export class Editor extends EventEmitter<TLEventMap> {
...
 
-					if (key !== 'color') {
-						const changes: TLShapePartial[] = []
-
-						for (const shape of shapesToUpdate) {
-							const currentShape = this.getShapeById(shape.id)
-							if (!currentShape) continue
-							const util = this.getShapeUtil(currentShape)
+					// TODO: find a way to sink this stuff into shape utils directly?
+					const changes: TLShapePartial[] = []
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/Editor.ts:7894]: should we chuck this into the `ShapeUtil` as `onSetStyle` or something?

File: packages/editor/src/lib/editor/shapes/arrow/components/DashedArrowComponent.tsx [hunk 7]
```diff
@@ -1,108 +0,0 @@
-import { CubicSpline2d, Polyline2d } from '@tldraw/primitives'
```
Comment (SomeHats) [inline packages/editor/src/lib/editor/shapes/arrow/components/DashedArrowComponent.tsx:1]: These are unused

File: packages/editor/src/lib/utils/SharedStylesMap.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			return b.type === 'mixed'
+		case 'shared':
+			return b.type === 'shared' && a.value === b.value
+		default:
+			throw exhaustiveSwitchError(a)
+	}
+}
+
+/** @public */
+export class ReadonlySharedStyleMap {
```
Comment (SomeHats) [inline packages/editor/src/lib/utils/SharedStylesMap.ts:22]: This is how we represent the replacement for `editor.props`

**Comment by SomeHats** [comment]
i'm going to land this as-is and implement the discussed tweaks as a follow-up

**Comment by steveruizok** [comment]
Looks like this one never created a corresponding PR on brivate?

**Pull request merged**